File: Ejercicios_propuestos/Propuesto1/Java_01/Rectangulo.java
package EjPropuesto.Java_01;

import java.util.Scanner;

public class Rectangulo {

    public double validarValor(String entrada) {
        try {
            double valor = Double.parseDouble(entrada.replace(",", "."));
            if (valor <= 0) {
                throw new IllegalArgumentException("El valor debe ser positivo.");
            }
            return valor;
        } catch (NumberFormatException e) {
            throw new NumberFormatException("Entrada inválida. Debe ser un número válido.");
        }
    }

    public double calcularArea(double base, double altura) {
        if (base <= 0 || altura <= 0) {
            throw new IllegalArgumentException("La base y altura deben ser positivas.");
        }
        return base * altura;
    }

    public void mostrarResultados(double base, double altura, double area) {
        System.out.println("\nResultados:");
        System.out.println("Base: " + base);
        System.out.println("Altura: " + altura);
        System.out.println("Área del rectángulo: " + area);
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        Rectangulo rectangulo = new Rectangulo();

        double base = 0;
        double altura = 0;

        while (true) {
            try {
                System.out.print("Ingresa la base del rectángulo: ");
                base = rectangulo.validarValor(scan.nextLine());
                break;
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
        }

        while (true) {
            try {
                System.out.print("Ingresa la altura del rectángulo: ");
                altura = rectangulo.validarValor(scan.nextLine());
                break;
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
        }

        double area = rectangulo.calcularArea(base, altura);
        rectangulo.mostrarResultados(base, altura, area);
        scan.close();
    }
}

File: Python/01/rectangulo.py
def obtener_valor(mensaje):
    while True:
        entrada = input(mensaje)
        try:
            valor = float(entrada.replace(",", "."))
            if valor <= 0:
                print("Por favor, ingresa un número positivo.")
            else:
                return valor
        except ValueError:
            print(
                "Entrada inválida. Asegúrate de ingresar un número (entero o decimal)."
            )


def calcular_area(base, altura):
    if not isinstance(base, (int, float)) or not isinstance(altura, (int, float)):
        raise TypeError("La base y la altura deben ser números.")
    if base <= 0 or altura <= 0:
        raise ValueError("La base y la altura deben ser valores positivos.")
    return base * altura


def calcular_area_rectangulo():
    print("Cálculo del área de un rectángulo\n")

    base = obtener_valor("Ingresa la base del rectángulo: ")
    altura = obtener_valor("Ingresa la altura del rectángulo: ")

    area = calcular_area(base, altura)

    print(f"\nResultados:")
    print(f"Base: {base}")
    print(f"Altura: {altura}")
    print(f"Área del rectángulo: {area}")


if __name__ == "__main__":
    calcular_area_rectangulo()
